### PR TITLE
Reimplement Deformable-detr

### PR DESCRIPTION
**Pull request by jshilong**

### CODE REVIEW

**Comment by ZwwWayne** [comment]
I suppose this PR relies on #4763

File: configs/deformable_detr/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+[ALGORITHM]
+
+We provide the config files for Deformable DETR: [Deformable DETR: Deformable Transformers for End-to-End Object Detection](https://arxiv.org/abs/2010.04159).
+
+```bibtex
+@article{zhu2020deformable,
+  title={Deformable DETR: Deformable Transformers for End-to-End Object Detection},
+  author={Zhu, Xizhou and Su, Weijie and Lu, Lewei and Li, Bin and Wang, Xiaogang and Dai, Jifeng},
+  journal={arXiv preprint arXiv:2010.04159},
```
Comment (ZwwWayne) [inline configs/deformable_detr/README.md:13]: Use ICLR bib

File: configs/deformable_detr/deformable_detr_r50_8x2_50e_coco.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            type='SinePositionalEncoding',
+            num_feats=128,
+            normalize=True,
+            offset=-0.5),
+        loss_cls=dict(
+            type='FocalLoss',
+            use_sigmoid=True,
+            gamma=2.0,
+            alpha=0.25,
+            loss_weight=2.0,
```
Comment (ZwwWayne) [inline configs/deformable_detr/deformable_detr_r50_8x2_50e_coco.py:72]: clean comma

File: configs/deformable_detr/deformable_detr_r50_8x2_50e_coco.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+# optimizer
+optimizer = dict(
+    type='AdamW',
+    lr=2e-4,
+    weight_decay=0.0001,
+    paramwise_cfg=dict(
+        custom_keys={
+            'backbone': dict(lr_mult=0.1),
+            'sampling_offsets': dict(lr_mult=0.1),
+            'reference_points': dict(lr_mult=0.1),
```
Comment (ZwwWayne) [inline configs/deformable_detr/deformable_detr_r50_8x2_50e_coco.py:167]: clean comma

File: configs/deformable_detr/deformable_detr_refine_r50_8x2_50e_coco.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+_base_ = ['deformable_detr_r50_8x2_50e_coco.py']
```
Comment (ZwwWayne) [inline configs/deformable_detr/deformable_detr_refine_r50_8x2_50e_coco.py:1]: directly use string rather than list[str]

File: configs/deformable_detr/deformable_detr_twostage_refine_r50_8x2_50e_coco.py [hunk 5]
```diff
@@ -0,0 +1,2 @@
+_base_ = ['deformable_detr_refine_r50_8x2_50e_coco.py']
```
Comment (ZwwWayne) [inline configs/deformable_detr/deformable_detr_twostage_refine_r50_8x2_50e_coco.py:1]: `_base_ = 'deformable_detr_refine_r50_8x2_50e_coco.py'`

File: mmdet/models/dense_heads/deformable_detr_head.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+            mlvl_masks.append(
+                F.interpolate(img_masks[None],
+                              size=feat.shape[-2:]).to(torch.bool).squeeze(0))
+            mlvl_positional_encodings.append(
+                self.positional_encoding(mlvl_masks[-1]))
+
+        query_embeds = None
+        if not self.as_two_stage:
+            query_embeds = self.query_embedding.weight
+        hs, init_reference, inter_references, \
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/deformable_detr_head.py:142]: what does `hs` stand for? hidden_state?

File: mmdet/models/dense_heads/detr_head.py [hunk 7]
```diff
@@ -369,6 +369,9 @@ def loss_single(self,
         cls_avg_factor = num_total_pos * 1.0 + \
             num_total_neg * self.bg_cls_weight
 
+        cls_avg_factor = reduce_mean(cls_scores.new_tensor([cls_avg_factor]))
+        cls_avg_factor = max(cls_avg_factor, 1)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/detr_head.py:373]: see line 375,
will reduce_mean affect the performance of DETR?

File: mmdet/models/detectors/detr.py [hunk 8]
```diff
@@ -32,6 +32,7 @@ def simple_test(self, img, img_metas, rescale=False):
                 corresponds to each class.
         """
         batch_size = len(img_metas)
+        # TOCO check this assetr
```
Comment (ZwwWayne) [inline mmdet/models/detectors/detr.py:35]: typo

File: mmdet/models/utils/transformer.py [hunk 9]
```diff
@@ -1,16 +1,28 @@
...
+                                         MultiScaleDeformableAttention,
                                          TransformerLayerSequence,
                                          build_transformer_layer_sequence)
 from mmcv.runner.base_module import BaseModule
+from torch.nn.init import normal_
 
 from mmdet.models.utils.builder import TRANSFORMER
 
 
+def inverse_sigmoid(x, eps=1e-5):
```
Comment (ZwwWayne) [inline mmdet/models/utils/transformer.py:18]: can we put it in another file?

File: mmdet/models/utils/transformer.py [hunk 10]
```diff
@@ -227,6 +239,413 @@ def forward(self, x, mask, query_embed, pos_embed):
...
+            query_embeds, tgt = torch.split(pos_trans_out, c, dim=2)
+        else:
+            query_embeds, tgt = torch.split(query_embed, c, dim=1)
+            query_embeds = query_embeds.unsqueeze(0).expand(bs, -1, -1)
+            tgt = tgt.unsqueeze(0).expand(bs, -1, -1)
+            reference_points = self.reference_points(query_embeds).sigmoid()
+            init_reference_out = reference_points
+
+        # decoder
+        tgt = tgt.permute(1, 0, 2)
```
Comment (ZwwWayne) [inline mmdet/models/utils/transformer.py:624]: tgt -> target

File: mmdet/models/utils/transformer.py (hunk 9 above)
Comment (jshilong) [reply mmdet/models/utils/transformer.py:18]: For exmaple?

File: configs/deformable_detr/deformable_detr_r50_8x2_50e_coco.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
+_base_ = [
```
Comment (ZwwWayne) [inline configs/deformable_detr/deformable_detr_r50_8x2_50e_coco.py:1]: 8x2 seems to be a default setting, we can omit it for simplicity.

File: mmdet/models/dense_heads/detr_head.py [hunk 12]
```diff
@@ -368,6 +368,11 @@ def loss_single(self,
         # construct weighted avg_factor to match with the official DETR repo
         cls_avg_factor = num_total_pos * 1.0 + \
             num_total_neg * self.bg_cls_weight
+        if self.__class__ is not DETRHead:
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/detr_head.py:371]: Can we simply add an option for that? For example `sync_cls_denominator`

File: mmdet/models/utils/transformer.py (hunk 9 above)
Comment (ZwwWayne) [inline mmdet/models/utils/transformer.py:18]: Add docstring, and import it in utils/__init__.py thus we can use it by `import inverse_sigmoid from mmdet.models.utils`.

File: configs/deformable_detr/deformable_detr_r50_8x2_50e_coco.py (hunk 11 above)
Comment (jshilong) [reply configs/deformable_detr/deformable_detr_r50_8x2_50e_coco.py:1]: should be 16*2.

File: mmdet/models/utils/transformer.py (hunk 9 above)
Comment (jshilong) [reply mmdet/models/utils/transformer.py:18]: this would cause circular dependency.

File: configs/deformable_detr/README.md [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+## Introduction
+
+<!-- [ALGORITHM] -->
+
+We provide the config files for Deformable DETR: [Deformable DETR: Deformable Transformers for End-to-End Object Detection](https://arxiv.org/abs/2010.04159).
+
+```
+@inproceedings{
+zhu2021deformable,
+title={Deformable {\{}DETR{\}}: Deformable Transformers for End-to-End Object Detection},
```
Comment (ZwwWayne) [inline configs/deformable_detr/README.md:12]: `Deformable {\{}DETR{\}}: Deformable` -> `{Deformable DETR: Deformable}`

File: configs/deformable_detr/README.md [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+| Backbone | Model | Lr schd  | box AP | Config | Download |
+|:------:|:--------:|:--------------:|:------:|:------:|:--------:|
+| R-50 | Deformable DETR  |50e  | 44.5 | [config]() | [model]() &#124; [log]() |
+| R-50 | + iterative bounding box refinement  |50e  | 46.1 | [config]() | [model]() &#124; [log]() |
+| R-50 | ++ two-stage Deformable DETR  |50e  | 46.8 | [config]() | [model]() &#124; [log]() |
+
+# NOTE
+
+1. The performance is unstable. `Deformable DETR` and `iterative bounding box refinement` may fluctuate about 0.3 mAP. `two-stage Deformable DETR` fluctuate about 0.2 mAP.
```
Comment (ZwwWayne) [inline configs/deformable_detr/README.md:30]: fluctuate -> fluctuates or may xxx

File: configs/deformable_detr/README.md [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+## Results and Models
+
+| Backbone | Model | Lr schd  | box AP | Config | Download |
+|:------:|:--------:|:--------------:|:------:|:------:|:--------:|
+| R-50 | Deformable DETR  |50e  | 44.5 | [config]() | [model]() &#124; [log]() |
+| R-50 | + iterative bounding box refinement  |50e  | 46.1 | [config]() | [model]() &#124; [log]() |
+| R-50 | ++ two-stage Deformable DETR  |50e  | 46.8 | [config]() | [model]() &#124; [log]() |
+
+# NOTE
+
```
Comment (ZwwWayne) [inline configs/deformable_detr/README.md:29]: Indicate the difference between ours and the original setting, e.g., different batch sizes or numbers of GPU cards?

File: mmdet/models/utils/transformer.py [hunk 16]
```diff
@@ -227,6 +239,436 @@ def forward(self, x, mask, query_embed, pos_embed):
...
+                - output_memory (Tensor): The input of decoder,  \
+                    has shape (bs, num_key, embed_dim).  num_key is \
+                    equal the number of points on feature map from \
+                    all levels.
+                - output_proposals (Tensor): The normalized proposal \
+                    after a inverse sigmoid, has shape \
+                    (bs, num_keys, 4).
+        """
+
+        N_, S_, C_ = memory.shape
```
Comment (ZwwWayne) [inline mmdet/models/utils/transformer.py:407]: Why cannot we directly use N, S, C, but need to use N_, S_, C_?

**Comment by ZwwWayne** [comment]
Last round, can be merged after comments are resolved.

File: configs/deformable_detr/README.md (hunk 15 above)
Comment (jshilong) [reply configs/deformable_detr/README.md:29]: This follows the setting in the original codebase.

**Review by ZwwWayne — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Comment by ZwwWayne** [comment]
resolve conflicts
